### ISSUE TITLE
Wrong datetime retrieved from database on DST day

### ISSUE THREAD

**Issue by fer22f** [tediousjs/tedious#625]
Hello!

I've had a single issue on the day 15, and found out the problem lies in tedious. The issue is repeatable, and predictable, any query in a database on the date 2017-10-15 **in my timezone** returns a datetime with an hour offset (+1). useUTC is disabled.

The issue lies in this line: https://github.com/tediousjs/tedious/blob/d2cf7186c9d8eda4e530f42d5be8072eb6481e85/src/value-parser.js#L519-L521

When the line 520 changes the date to the faulty DST date 2017-10-15, it appropriately changes the time to 01:00:00, since 00:00:00 doesn't **exist**. So, in line 512, the Date object **sums** the 1 hour to the milliseconds from the database, which results in an offset.

The way to fix this is by removing the 1 hour difference from the Date object at that place, which isn't feasible directly because 00:00:00 doesn't exist. So we have to derive the current milliseconds offset in some way like calculating `getHours() * 60 * 60 * 1000` and then subtracting that from `milliseconds`.

I can submit a pull request later, but I really don't know how I would approach writing an unit test for this, since it seems dependent on the Date implementation and the timezone of the user.

**Comment by fer22f** [comment]
While making the tests, I figured out there was more to this, DST doesn't work the same way here in Brazil as it does it the US, and that's one key problem.

On Brazil, 00:00:00 doesn't exist. Only 01:00:00 forward exists in the DST day.
On the US, the change occurs at 02:00:00, so the issue doesn't happen at all, since 00:00:00 exists.

My proposed solution actually still works (the offset is from 00:00:00, not from nothing else), but I have to take this into account when making proper unit tests. `timezone-mock` unfortunately doesn't have Brazil's locale, I'll be contacting them about it.

**Comment by arthurschreiber** [comment]
@fer22f Thanks for your bug report! ❤️ 

According to section [`2.2.5.5.1.8`](https://msdn.microsoft.com/en-us/library/ee780895.aspx) of the `MS-TDS` specification, the calculation as done by Tedious here should be correct.

What might be incorrect here is how we create the `Date` object and/or how we modify it to reflect the data we received.

Could you provide more information around this? What are the exact values the `readInt32LE` and the `readUInt32LE` calls return here? Also, what is your locale (both the server's as well as the client's locale).

Providing this information will help  us track down whatever is going wrong here. 😄

**Comment by fer22f** [comment]
Yea, the date created by the library is always correct, the only problem is the resulting `Date` object. The problem from the spec being:

`One 2-byte unsigned integer that represents the number of minutes elapsed since 12 AM that day.`

Which doesn't happen on my case, since setting the date to the DST date makes 12AM suddenly expire (it jumps to 1AM).

But here it is, I passed a list of days and milliseconds to this function and here are the results:

```javascript
function fn([{ days, milli }]) {
  const value = new Date(1900, 0, 1);
  value.setDate(value.getDate() + days);
  value.setMilliseconds(value.getMilliseconds() + milli );
  return { date: value, days, milli: milliseconds, value: value.toLocaleString() }
}

// Input/Output:
[
  { date: 2017-10-14T03:00:00.000Z, days: 43020, milli: 0, value: '2017-10-14 00:00:00' },
  { date: 2017-10-14T03:30:00.000Z, days: 43020, milli: 1800000, value: '2017-10-14 00:30:00' },
  { date: 2017-10-14T04:05:00.000Z, days: 43020, milli: 3900000, value: '2017-10-14 01:05:00' },
  { date: 2017-10-14T10:23:00.000Z, days: 43020, milli: 26580000, value: '2017-10-14 07:23:00' },
  { date: 2017-10-14T15:30:00.000Z, days: 43020, milli: 45000000, value: '2017-10-14 12:30:00' },
  { date: 2017-10-14T15:38:00.000Z, days: 43020, milli: 45480000, value: '2017-10-14 12:38:00' },
  { date: 2017-10-14T20:37:00.000Z, days: 43020, milli: 63420000, value: '2017-10-14 17:37:00' },
  { date: 2017-10-14T21:03:00.000Z, days: 43020, milli: 64980000, value: '2017-10-14 18:03:00' },
  { date: 2017-10-14T21:47:00.000Z, days: 43020, milli: 67620000, value: '2017-10-14 18:47:00' },
  { date: 2017-10-14T23:19:00.000Z, days: 43020, milli: 73140000, value: '2017-10-14 20:19:00' },
  { date: 2017-10-15T03:00:00.000Z, days: 43021, milli: 0, value: '2017-10-15 01:00:00' },
  { date: 2017-10-15T03:30:00.000Z, days: 43021, milli: 1800000, value: '2017-10-15 01:30:00' },
  { date: 2017-10-15T04:05:00.000Z, days: 43021, milli: 3900000, value: '2017-10-15 02:05:00' },
  { date: 2017-10-15T10:23:00.000Z, days: 43021, milli: 26580000, value: '2017-10-15 08:23:00' },
  { date: 2017-10-15T15:30:00.000Z, days: 43021, milli: 45000000, value: '2017-10-15 13:30:00' },
  { date: 2017-10-15T20:37:00.000Z, days: 43021, milli: 63420000, value: '2017-10-15 18:37:00' },
  { date: 2017-10-15T21:03:00.000Z, days: 43021, milli: 64980000, value: '2017-10-15 19:03:00' },
  { date: 2017-10-15T21:47:00.000Z, days: 43021, milli: 67620000, value: '2017-10-15 19:47:00' },
  { date: 2017-10-15T23:19:00.000Z, days: 43021, milli: 73140000, value: '2017-10-15 21:19:00' },
  { date: 2017-10-16T02:00:00.000Z, days: 43022, milli: 0, value: '2017-10-16 00:00:00' },
  { date: 2017-10-16T02:30:00.000Z, days: 43022, milli: 1800000, value: '2017-10-16 00:30:00' },
  { date: 2017-10-16T03:05:00.000Z, days: 43022, milli: 3900000, value: '2017-10-16 01:05:00' },
  { date: 2017-10-16T09:23:00.000Z, days: 43022, milli: 26580000, value: '2017-10-16 07:23:00' }
]
```

The issue happens exactly in day 15. You can see when milli is 3900000 that it outputs 02:05, while in the other days, it outputs 01:05, correctly. One glaring issue is that effectively, milli 0 and milli 1800000 are contained within the "non-existent" hour and should not exist at all. Here is how JavaScript works that out:

```javascript
function fn({ days, milli }) {
  const value = new Date(1900, 0, 1 + days, 0, 0, 0, milli);
  return { date: value, days, milli: milli, value: value.toLocaleString() }
}

[
  { date: 2017-10-14T03:00:00.000Z, days: 43020, milli: 0, value: '2017-10-14 00:00:00' },
  { date: 2017-10-14T03:30:00.000Z, days: 43020, milli: 1800000, value: '2017-10-14 00:30:00' },
  { date: 2017-10-14T04:05:00.000Z, days: 43020, milli: 3900000, value: '2017-10-14 01:05:00' },
  { date: 2017-10-14T10:23:00.000Z, days: 43020, milli: 26580000, value: '2017-10-14 07:23:00' },
  { date: 2017-10-14T15:30:00.000Z, days: 43020, milli: 45000000, value: '2017-10-14 12:30:00' },
  { date: 2017-10-14T15:38:00.000Z, days: 43020, milli: 45480000, value: '2017-10-14 12:38:00' },
  { date: 2017-10-14T20:37:00.000Z, days: 43020, milli: 63420000, value: '2017-10-14 17:37:00' },
  { date: 2017-10-14T21:03:00.000Z, days: 43020, milli: 64980000, value: '2017-10-14 18:03:00' },
  { date: 2017-10-14T21:47:00.000Z, days: 43020, milli: 67620000, value: '2017-10-14 18:47:00' },
  { date: 2017-10-14T23:19:00.000Z, days: 43020, milli: 73140000, value: '2017-10-14 20:19:00' },
  { date: 2017-10-15T03:00:00.000Z, days: 43021, milli: 0, value: '2017-10-15 01:00:00' },
  { date: 2017-10-15T03:30:00.000Z, days: 43021, milli: 1800000, value: '2017-10-15 01:30:00' },
  { date: 2017-10-15T03:05:00.000Z, days: 43021, milli: 3900000, value: '2017-10-15 01:05:00' },
  { date: 2017-10-15T09:23:00.000Z, days: 43021, milli: 26580000, value: '2017-10-15 07:23:00' },
  { date: 2017-10-15T14:30:00.000Z, days: 43021, milli: 45000000, value: '2017-10-15 12:30:00' },
  { date: 2017-10-15T19:37:00.000Z, days: 43021, milli: 63420000, value: '2017-10-15 17:37:00' },
  { date: 2017-10-15T20:03:00.000Z, days: 43021, milli: 64980000, value: '2017-10-15 18:03:00' },
  { date: 2017-10-15T20:47:00.000Z, days: 43021, milli: 67620000, value: '2017-10-15 18:47:00' },
  { date: 2017-10-15T22:19:00.000Z, days: 43021, milli: 73140000, value: '2017-10-15 20:19:00' },
  { date: 2017-10-16T02:00:00.000Z, days: 43022, milli: 0, value: '2017-10-16 00:00:00' },
  { date: 2017-10-16T02:30:00.000Z, days: 43022, milli: 1800000, value: '2017-10-16 00:30:00' },
  { date: 2017-10-16T03:05:00.000Z, days: 43022, milli: 3900000, value: '2017-10-16 01:05:00' },
  { date: 2017-10-16T09:23:00.000Z, days: 43022, milli: 26580000, value: '2017-10-16 07:23:00' }
]
```

Using JavaScript in our favor, the dates are set correctly (I really don't know what the output should be for 1800000 to be frank). And this is the optimal solution (not the one I talked before about offsets and all).

Both the client and SQL Server are in the same machine, both in locale [0416](https://www.microsoft.com/resources/msdn/goglobal/default.mspx?submitted=0416&OS=Windows%207). The default login language is Brazilian in SQL Server. Our timezone is both UTC-3 (Standard) and UTC-2 (DST), and the timezone begins working at 12AM (not 2AM), hence the issue.

**Comment by arthurschreiber** [comment]
Nice, awesome work digging into this! ❤️ 😍 

The fix for this should be pretty simple to prepare. Do you want to take a stab at it?

**Comment by arthurschreiber** [comment]
Note: We should also be using `Date.UTC` when creating the UTC date.

**Comment by fer22f** [comment]
Sure thing! I will submit a PR shortly, but the unit tests for this seem... hard to come by, and sadly that mock library doesn't have this specific use case. But I will submit the PR without them, for now.

**Comment by arthurschreiber** [comment]
This got closed by #628.

**Comment by arthurschreiber** [comment]
@fer22f The fix for this was released in `tedious@2.1.2`. It's currently available via the `next` release tag - you can install it via `npm install tedious@next` or `npm install tedious@2.1.2`.

Please let me know if this version works for you, and I'll bump it to `latest` to mark it as stable! 😄
